ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -0,0 +1,9955 @@
+{
+	ZEND_ASSERT(Z_MODE(src) == IS_REG);
+	ZEND_ASSERT(Z_MODE(dst) == IS_MEM_ZVAL);
+
+	if ((info & MAY_BE_ANY) == MAY_BE_LONG) {
+		|	SET_ZVAL_LVAL dst, Ra(Z_REG(src))
+		if (set_type) {
+			|	SET_ZVAL_TYPE_INFO dst, IS_LONG
+		}
+	} else if ((info & MAY_BE_ANY) == MAY_BE_DOUBLE) {
+		|	.if SSE
+		|		SSE_SET_ZVAL_DVAL dst, Z_REG(src)
+		|	.else
+		||		ZEND_ASSERT(0);
+		|	.endif
+		if (set_type) {
+			|	SET_ZVAL_TYPE_INFO dst, IS_DOUBLE
+		}
+	} else {
+		ZEND_ASSERT(0);
+	}
+	return 1;
+}
+
+static int zend_jit_load_reg(dasm_State **Dst, zend_jit_addr src, zend_jit_addr dst, uint32_t info)
+{
+	ZEND_ASSERT(Z_MODE(src) == IS_MEM_ZVAL);
+	ZEND_ASSERT(Z_MODE(dst) == IS_REG);
+
+	if ((info & MAY_BE_ANY) == MAY_BE_LONG) {
+		|	GET_ZVAL_LVAL Z_REG(dst), src
+	} else if ((info & MAY_BE_ANY) == MAY_BE_DOUBLE) {
+		|	.if SSE
+		|		SSE_GET_ZVAL_DVAL Z_REG(dst), src
+		|	.else
+		||		ZEND_ASSERT(0);
+		|	.endif
+	} else {
+		ZEND_ASSERT(0);
+	}
+	return 1;
+}
+
+static int zend_jit_store_ssa_var(dasm_State **Dst, zend_ssa *ssa, int var, zend_reg reg)
+{
+	zend_jit_addr src = ZEND_ADDR_REG(reg);
+	zend_jit_addr dst = ZEND_ADDR_MEM_ZVAL(ZREG_FP, (uint32_t)(uintptr_t)ZEND_CALL_VAR_NUM(NULL, ssa->vars[var].var));
+	uint32_t info = ssa->var_info[var].type;
+
+	return zend_jit_spill_store(Dst, src, dst, info, 1);
+}
+
+static int zend_jit_store_ssa_var_if_necessary(dasm_State **Dst, zend_ssa *ssa, zend_lifetime_interval **ra, zend_jit_addr src, int var, int old_var)
+{
+	if (Z_MODE(src) == IS_REG && ra[var] && ra[var]->store) {
+		zend_jit_addr dst = ZEND_ADDR_MEM_ZVAL(ZREG_FP, (uint32_t)(uintptr_t)ZEND_CALL_VAR_NUM(NULL, ssa->vars[var].var));
+		uint32_t info = ssa->var_info[var].type;
+		zend_bool set_type = 1;
+
+		if (old_var >= 0) {
+			uint32_t old_info = ssa->var_info[old_var].type;
+			if ((info & (MAY_BE_ANY|MAY_BE_REF|MAY_BE_UNDEF)) ==
+			    (old_info & (MAY_BE_ANY|MAY_BE_REF|MAY_BE_UNDEF))) {
+				if (!ra[old_var] || ra[old_var]->load || ra[old_var]->store) {
+					set_type = 0;
+				}
+			}
+		}
+		return zend_jit_spill_store(Dst, src, dst, info, set_type);
+	}
+	return 1;
+}
+
+static int zend_jit_load_ssa_var(dasm_State **Dst, zend_ssa *ssa, int var, zend_reg reg)
+{
+	zend_jit_addr src = ZEND_ADDR_MEM_ZVAL(ZREG_FP, (uint32_t)(uintptr_t)ZEND_CALL_VAR_NUM(NULL, ssa->vars[var].var));
+	zend_jit_addr dst = ZEND_ADDR_REG(reg);
+	uint32_t info = ssa->var_info[var].type;
+
+	return zend_jit_load_reg(Dst, src, dst, info);
+}
+
+static int zend_jit_update_regs(dasm_State **Dst, zend_jit_addr src, zend_jit_addr dst, uint32_t info, zend_lifetime_interval *src_ival)
+{
+	if (src != dst) {
+		if (Z_MODE(src) == IS_REG) {
+			if (Z_MODE(dst) == IS_REG) {
+				if ((info & MAY_BE_ANY) == MAY_BE_LONG) {
+					|	mov Ra(Z_REG(dst)), Ra(Z_REG(src))
+				} else if ((info & MAY_BE_ANY) == MAY_BE_DOUBLE) {
+					|	.if SSE
+					|		SSE_AVX_INS movsd, vmovaps, xmm(Z_REG(dst)-ZREG_XMM0), xmm(Z_REG(src)-ZREG_XMM0)
+					|	.else
+					||		ZE
```

Comment:
Is this separation necessary (even in the assing op case)? add_function etc should separate themselves as necessary.

Reply:
I didn't get. What separation?

Reply:
Something is broken with the display here. I meant the SEPARATE_ZVAL_NOREF in zend_jit_math_helper.